README.md
# SPARQL Faceter

A paper describing SPARQL Faceter can be found [here](http://seco.cs.aalto.fi/publications/2016/koho-et-al-sparql-faceter.pdf),
although as of version 1.0.0 the implementation is quite different.
Most importantly, rather than there being one SPARQL query to get the state of all the facets,
each facet makes its own query.

## Installation

Install via Bower:

`bower install sparql-faceter`

Include `seco.facetedSearch` in your module dependencies:

```
angular.module('myApp', ['seco.facetedSearch'])
```

Include the script in your HTML:

```
<script src="bower_components/sparql-faceter/dist/semantic-faceted-search.js></script>
```

And additionally the dependencies defined in [bower.json](bower.json) before the above script.

You can also include the stylesheet in the header:

```
<link rel="stylesheet" href="bower_components/angular-semantic-faceted-search/dist/semantic-faceted-search.css">
```

Then you're good to go!

## Configuration and Usage

See the [documentation](http://semanticcomputing.github.io/angular-semantic-faceted-search/#).

Setup in the controller:

```
var vm = this;

// Define facets
vm.facets = {
    // Text facet
    name: {
        name: 'Name',
        facetId: 'name',
        predicate: '<http://www.w3.org/2004/02/skos/core#prefLabel>',
        enabled: true
    },
    // Date facet
    deathDate: {
        name: 'Time of Death',
        facetId: 'death',
        startPredicate: '<http://ldf.fi/schema/narc-menehtyneet1939-45/kuolinaika>',
        endPredicate: '<http://ldf.fi/schema/narc-menehtyneet1939-45/kuolinaika>',
        min: '1939-10-01',
        max: '1989-12-31',
        enabled: true
    },
    // Basic facet
    profession: {
        name: 'Ammatti',
        facetId: 'occupation',
        predicate: '<http://ldf.fi/schema/narc-menehtyneet1939-45/ammatti>',
        enabled: true
    },
    // Basic facet with property path
    source: {
        name: 'Source',
        facetId: 'source',
        predicate: '^<http://www.cidoc-crm.org/cidoc-crm/P70i_is_documented_in>/<http://purl.org/dc/elements/1.1/source>',
        enabled: true
    },
    // Basic facet with labels in another service.
    birthMunicipality: {
        name: 'Birth Municipality',
        services: ['<http://ldf.fi/pnr/sparql>'],
        facetId: 'birthplace',
        predicate: '<http://ldf.fi/schema/narc-menehtyneet1939-45/synnyinkunta>',
        enabled: false
    },
    // Hierarchical facet
    rank: {
        name: 'Rank',
        facetId: 'rank',
        predicate: '<http://ldf.fi/schema/narc-menehtyneet1939-45/sotilasarvo>',
        hierarchy: '<http://purl.org/dc/terms/isPartOf>',
        enabled: true,
        depth: 3
    }
};

// Define common options
vm.options = {
    scope: $scope,
    endpointUrl: 'http://ldf.fi/warsa/sparql',
    rdfClass: '<http://ldf.fi/schema/narc-menehtyneet1939-45/DeathRecord>',
    constraint: '?id skos:prefLabel ?name .',
    preferredLang : 'fi'
};

// Define a function that handles updates.
// 'dataService' is some service that fetches results based on the facet selections.
function updateResults(event, facetState) {
    dataService.getResults(facetState.constraints).then(function(results) {
        vm.results = results;
    }
}

// Listen for the update event
$scope.$on('sf-facet-constraints', updateResults);

// Listen for initial state
var initListener = $scope.$on('sf-initial-constraints', function(event, state) {
    updateResults(event, state);
    // Only listen once for the init event
    initListener();
});

// Initialize the facet handler:
vm.handler = new FacetHandler(vm.options);
```

Then, in the template:

```
<seco-text-facet
  data-options="vm.facets.name">
</seco-text-facet>
<seco-timespan-facet
  data-options="vm.facets.deathDate">
</seco-timespan-facet>
<seco-basic-facet
  data-options="vm.facets.source">
</seco-basic-facet>
<seco-basic-facet
  data-options="vm.facets.profession">
</seco-basic-facet>
<seco-basic-facet
  data-options="vm.facets.birthMunicipality">
</seco-basic-facet>
<seco-basic-facet
  data-options="vm.facets.principalAbode">
</seco-basic-facet>
<seco-hierarchy-facet
  data-options="vm.facets.rank">
</seco-hierarchy-facet>
```

## Examples
[Simplistic demo using DBpedia](http://semanticcomputing.github.io/sparql-faceter-dbpedia-demo/#)
(the code in the [repository](https://github.com/SemanticComputing/sparql-faceter-dbpedia-demo) contains helpful comments)

The [WarSampo casualties demo](https://github.com/SemanticComputing/casualties-demo) uses SPARQL Faceter.

You can also see the tool in action in the [WarSampo photograph perspective](http://www.sotasampo.fi/en/photographs).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/),
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

### Added

- Configuration option for displaying facets as pie charts. If the option `chart` is truthy,
  the user can click a button next to the disable facet button to see and interact with the
  facet as a pie chart. The option is applicable for
  [BasicFacet](http://semanticcomputing.github.io/angular-semantic-faceted-search/#/api/seco.facetedSearch.directive:secoBasicFacet)
  and [CheckboxFacet](http://semanticcomputing.github.io/angular-semantic-faceted-search/#/api/seco.facetedSearch.directive:secoCheckboxFacet).
  It also works with
  [HierachyFacet](http://semanticcomputing.github.io/angular-semantic-faceted-search/#/api/seco.facetedSearch.directive:secoHierarchyFacet),
  but the hierarchy is not visualized, so usage with that facet is not recommended.
  The option is disabled by default.

## [1.10.0] - 2018-03-13

### Changed

- Cosmetic changes:
    - Facet disable button is now more subtle.
    - The enable/disable button is consistently placed.
    - Loading spinner is smaller.
- Refactored facet templates (this has some DOM/CSS implications).

## [1.9.1] - 2017-11-28

### Fixed

- Disabling and enabling a facet now works, instead of leaving the facet
  unresponsive.

## [1.9.0] - 2017-11-27

### Changed

- When an error occurs, don't try to display the error message to the user.

## [1.8.1] - 2017-10-16

### Changed

- Optimize hierarchy facet constraint.

## [1.8.0] - 2017-09-14

### Changed

- Add configuration option for HTTP headers.

## [1.7.2] - 2017-08-07

### Fixed

- Make the `priority` option work.

## [1.7.1] - 2017-08-03

### Changed

- Hierarchical facet now supports the `specifier` option.
- Hierarchical facet now displays only one dash per level in the hierarchy.

## [1.7.0] - 2017-07-26

### Changed

- Hierarchical facet now supports multiple levels in the hierarchy.
  The `classes` option is no longer needed.

## [1.6.1] - 2017-07-24

### Changed

- Update angular-spinner dependency.

## [1.6.0] - 2017-07-14

### Changed

- Make release version safe for minification.

## [1.5.4] - 2017-07-11

### Fixed

- Hierarchical facet now always shows all values.

## [1.5.3] - 2017-05-02

### Fixed

- FacetResultHandler now works without paging.

## [1.5.2] - 2017-03-27

### Fixed

- Queries to additional endpoints now return labels for values.

## [1.5.1] - 2017-03-24

### Fixed

- Timespan facet does not break anymore if its query returns undefined dates.

## [1.5.0] - 2017-03-15

### Added

- Checkbox facet (secoCheckboxFacet)

### Changed

- Simplified facet state update
- BasicFacet no longer does federated queries - a separate query for each
  service is produced instead.
- Multiple languages can now be given as the `preferredLang`.

## [1.4.1] - 2017-01-16

### Changed
- Values for the timespan facet are now cast as xsd:date in SPARQL

### Fixed
- Fix timespan facet date limits when selection changes

## [1.4.0] - 2017-01-11

### Changed
- Selecting a value in a basic facet does not update the list of selections anymore.
  This way the user can change the selection without clearing the selection first.
- Facet selections should now work with all literal datatypes.

## [1.3.2] - 2016-12-15

### Fixed
- Timespan facet selections restrictions fixed.

## [1.3.1] - 2016-11-15

### Fixed
- Sanitize the user's query in secoJenaTextFacet to avoid syntax errors from the
  SPARQL endpoint.

## [1.3.0] - 2016-11-15

### Changed
- secoJenaTextFacet allows for more user control in the search.
  I.e. it does not modify the user input except for excaping quotes
  (and removing them if they are unbalanced)
- secoJenaTextFacet default priority is now 0 (instead of 10).

## [1.2.1] - 2016-11-15

### Added
- `limit` and `graph` configuration options for secoJenaTextFacet.

## [1.2.0] - 2016-11-14

### Added
- [secoJenaTextFacet](http://semanticcomputing.github.io/angular-semantic-faceted-search/#/api/seco.facetedSearch.directive:secoJenaTextFacet)
- `priority` configuration variable to provide a way to sort the facet constraints (see the [documentation][api]).


## [1.1.1] - 2016-11-04

### Fixed
- Dates are hard. Dates should now appear the same in queries, URL params,
  and in the date picker, regardless of the user's timezone.

## [1.1.0] - 2016-11-03

### Changed
- Timespan facet now works as an actual facet by querying for minimum and maximum
  available dates, and adjusts the datepicker max and min dates accordingly.

## [1.0.7] - 2016-11-03

### Fixed
- Fix setting the timespan face's initial values.
- Fix an issue with the timespan facet where the selected date could be different
  from the one actually used in the query.

## [1.0.6] - 2016-11-02

### Added
- Changelog

### Changed
- UI Bootstrap dependency updated to ^2.2.0

### Fixed
- Fix the timespan facet, and its documentation

[Unreleased]: https://github.com/SemanticComputing/angular-semantic-faceted-search/compare/1.10.0...HEAD
[1.10.0]: https://github.com/SemanticComputing/angular-semantic-faceted-search/compare/1.9.1...1.10.0
[1.9.1]: https://github.com/SemanticComputing/angular-semantic-faceted-search/compare/1.9.0...1.9.1
[1.9.0]: https://github.com/SemanticComputing/angular-semantic-faceted-search/compare/1.8.1...1.9.0
[1.8.1]: https://github.com/SemanticComputing/angular-semantic-faceted-search/compare/1.8.0...1.8.1
[1.8.0]: https://github.com/SemanticComputing/angular-semantic-faceted-search/compare/1.7.1...1.8.0
[1.7.1]: https://github.com/SemanticComputing/angular-semantic-faceted-search/compare/1.7.0...1.7.1
[1.7.0]: https://github.com/SemanticComputing/angular-semantic-faceted-search/compare/1.6.1...1.7.0
[1.6.1]: https://github.com/SemanticComputing/angular-semantic-faceted-search/compare/1.6.0...1.6.1

facetedsearch.js
/*
 * Semantic faceted search
 *
 */

(function() {

    'use strict';

    /* eslint-disable angular/no-service-method */
    angular.module('facetApp')

    /*
    * Result handler service.
    */
    .factory( 'Results', function( RESULTS_PER_PAGE, PAGES_PER_QUERY, AdvancedSparqlService,
                personMapperService, FacetSelectionFormatter ) {
        return function( endpointUrl, facets ) {

            var formatter = new FacetSelectionFormatter(facets);
            var endpoint = new AdvancedSparqlService(endpointUrl, personMapperService);

            this.getResults = getResults;

            function getResults(facetSelections, query, resultSetQry) {
                return endpoint.getObjects(
                    query.replace('<FACET_SELECTIONS>', formatter.parseFacetSelections(facetSelections)),
                    RESULTS_PER_PAGE,
                    resultSetQry.replace('<FACET_SELECTIONS>', formatter.parseFacetSelections(facetSelections)),
                    PAGES_PER_QUERY);
            }
        };
    })

    /*
     * Casualty service
     */
    .service( 'casualtyService', function( $q, SparqlService, Results ) {
        var endpointUrl = 'http://ldf.fi/warsa/sparql';

        var facets = {
            '<http://www.w3.org/2004/02/skos/core#prefLabel>': { name: 'Nimi', type: 'text' },
            '<kuolinaika>' : {
                name: 'Kuolinaikarange',
                type: 'timespan',
                start: '<http://ldf.fi/schema/narc-menehtyneet1939-45/kuolinaika>',
                end: '<http://ldf.fi/schema/narc-menehtyneet1939-45/kuolinaika>',
                min: '1939-10-01',
                max: '1989-12-31'
            },
            '<http://ldf.fi/schema/narc-menehtyneet1939-45/ammatti>': { name: 'Ammatti' },
            '<http://ldf.fi/schema/narc-menehtyneet1939-45/asuinkunta>': { name: 'Asuinkunta' },
            '<http://ldf.fi/schema/narc-menehtyneet1939-45/kansalaisuus>': { name: 'Kansalaisuus' },
            '<http://ldf.fi/schema/narc-menehtyneet1939-45/kuolinkunta>': { name: 'Kuolinkunta' },
            '<http://ldf.fi/schema/narc-menehtyneet1939-45/kuolinaika>': { name: 'Kuolinaika' },
            '<http://ldf.fi/schema/narc-menehtyneet1939-45/osasto>': { name: 'Joukko-osasto' },
            '<http://ldf.fi/schema/narc-menehtyneet1939-45/sukupuoli>': { name: 'Sukupuoli' },
            '<http://ldf.fi/schema/narc-menehtyneet1939-45/siviilisaeaety>': { name: 'Siviilisääty' }
        };

        var resultHandler = new Results(endpointUrl, facets);

        var properties = {
            '?name': '',
            '?occupation': '<http://ldf.fi/schema/narc-menehtyneet1939-45/ammatti>',
            '?marital_status': '',
            '?death_municipality': '',
            '?tod': '',
            '?rank': '',
            '?unit': '',
            '?casualty_class': ''
        };

        var facetOptions = {
            endpointUrl: endpointUrl,
            graph : '<http://ldf.fi/narc-menehtyneet1939-45/>',
            rdfClass: '<http://xmlns.com/foaf/0.1/Person>',
            preferredLang : 'fi'
        };

        var prefixes = '' +
            ' PREFIX skos: <http://www.w3.org/2004/02/skos/core#>' +
            ' PREFIX wgs84: <http://www.w3.org/2003/01/geo/wgs84_pos#>' +
            ' PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>' +
            ' PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>' +
            ' PREFIX foaf: <http://xmlns.com/foaf/0.1/>' +
            ' PREFIX owl:  <http://www.w3.org/2002/07/owl#>' +
            ' PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>' +
            ' PREFIX georss: <http://www.georss.org/georss/>' +
            ' PREFIX text: <http://jena.apache.org/text#>' +
            ' PREFIX m: <http://ldf.fi/sotasampo/narc/menehtyneet/>' +
            ' PREFIX m_schema: <http://ldf.fi/schema/narc-menehtyneet1939-45/>';

        var resultSet = '' +
            '     SELECT ?s ?id ?name { ' +
            '       GRAPH <http://ldf.fi/narc-menehtyneet1939-45/> {' +
            '         <FACET_SELECTIONS> ' +
            '         ?s a foaf:Person .' +
            '         ?s skos:prefLabel ?name .' +
            '         BIND(?s AS ?id) ' +
            '       } ' +
            '     } ORDER BY ?name ' +
            '     <PAGE> ';

        var resultSetQry = prefixes + resultSet;

        var query = prefixes +
            ' SELECT ?id ?s <PROPERTIES> ' +
            ' WHERE {' +
            '   { ' +
            '     <RESULTSET> ' +
            '   } ' +
            ' GRAPH <http://ldf.fi/narc-menehtyneet1939-45/> {' +

            ' OPTIONAL {' +
            ' ?s m_schema:siviilisaeaety ?siviilisaeaetyuri .' +
            ' ?siviilisaeaetyuri skos:prefLabel ?marital_status . }' +
//            ' OPTIONAL { ?s m_schema:sukupuoli ?sukupuoliuri .' +
//            ' ?sukupuoliuri skos:prefLabel ?sukupuoli . }' +
//            ' OPTIONAL { ?s m_schema:lasten_lukumaeaerae ?lasten_lukumaeaerae . }' +
//            ' OPTIONAL { ?s m_schema:kansalaisuus ?kansalaisuusuri .' +
//            ' ?kansalaisuusuri skos:prefLabel ?kansalaisuus . }' +
//            ' OPTIONAL { ?s m_schema:aeidinkieli ?aeidinkieliuri .' +
//            ' ?aeidinkieliuri skos:prefLabel ?aeidinkieli . }' +
            ' OPTIONAL { ?s m_schema:menehtymisluokka ?menehtymisluokkauri .' +
            ' ?menehtymisluokkauri skos:prefLabel ?casualty_class . }' +
            ' OPTIONAL { ?s m_schema:kuolinkunta ?kuolinkunta_uri .' +
            ' OPTIONAL {' +
            ' 	GRAPH <http://ldf.fi/warsa/places/municipalities> {' +
            ' 		?kuolinkunta_uri skos:prefLabel ?kuolinkunta_warsa .' +
            ' 	}' +
            ' } OPTIONAL {' +
            ' 	?kuolinkunta_uri skos:prefLabel ?kuolinkunta_narc .' +
            ' }' +
            ' }' +
            ' OPTIONAL { ?s m_schema:kuolinaika ?tod .' +

//                    ' BIND(CONCAT("<a href=\"http://www.sotasampo.fi/times/page?uri=http://ldf.fi/warsa/events/times/time_",STR(?kuolinaika),"-",STR(?kuolinaika),"\">",STR(?kuolinaika),"</a>") AS ?kuolinpaiva) .' +
            ' }' +
            ' OPTIONAL { ?s m_schema:ammatti ?occupation . }' +
//                    ' OPTIONAL { ?s m_schema:kuolinpaikka ?kuolinpaikka . }' +
            ' OPTIONAL { ?s m_schema:sotilasarvo ?arvouri .' +
            '   GRAPH <http://ldf.fi/warsa/actors/actor_types> {' +
            '     ?arvouri skos:prefLabel ?rank  .' +
            '   }' +
            ' }' +
            ' OPTIONAL { ?s m_schema:osasto ?osastouri .' +
            '   GRAPH <http://ldf.fi/warsa/actors> {' +
            '     ?osastouri skos:prefLabel ?unit  .' +
            '   }' +
            ' }' +

            ' }' +
                ' BIND(COALESCE(?kuolinkunta_warsa, ?kuolinkunta_narc) as ?death_municipality)' +
            ' } ORDER BY ?name ';

        query = query.replace(/<RESULTSET>/g, resultSet);
        query = query.replace(/<PROPERTIES>/g, Object.keys( properties ).join(' '));

        this.getResults = getResults;
        this.getFacets = getFacets;
        this.getFacetOptions = getFacetOptions;

        function getResults(facetSelections) {
            return resultHandler.getResults(facetSelections, query, resultSetQry);
        }

        function getFacets() {
            return facets;
        }

        function getFacetOptions() {
            return facetOptions;
        }
    })

    /*
    * Controller for the results view.
    */
    .controller( 'MainController', function ( $q, _, RESULTS_PER_PAGE,
                casualtyService, NgTableParams ) {
        var vm = this;

        vm.facets = casualtyService.getFacets();
        vm.facetOptions = casualtyService.getFacetOptions();
        vm.facetOptions.updateResults = updateResults;

        vm.disableFacets = disableFacets;

        function disableFacets() {
            return vm.isLoadingResults;
        }

        function initializeTable() {
            vm.tableParams = new NgTableParams(
                {
                    count: RESULTS_PER_PAGE
                },
                {
                    getData: getData
                }
            );
        }

        function getData($defer, params) {
            vm.isLoadingResults = true;

            vm.pager.getPage(params.page() - 1, params.count())
            .then( function( page ) {
                $defer.resolve( page );
                vm.pager.getTotalCount().then(function(count) {
                    vm.tableParams.total( count );
                }).then(function() {
                    vm.isLoadingResults = false;
                });
            });
        }

        function updateResults( facetSelections ) {
            vm.isLoadingResults = true;

            casualtyService.getResults( facetSelections )
            .then( function ( pager ) {
                vm.pager = pager;
                if (vm.tableParams) {
                    vm.tableParams.page(1);
                    vm.tableParams.reload();
                } else {
                    initializeTable();
                }
            });
        }
    });
})();
